cpp/src/arrow/device.h

```diff
@@ -98,6 +100,64 @@ class ARROW_EXPORT Device : public std::enable_shared_from_this<Device>,
   /// \brief Return the DeviceAllocationType of this device
   virtual DeviceAllocationType device_type() const = 0;
 
+  /// \brief EXPERIMENTAL: An object that provides event/stream sync primitives
+  ///
+  ///
+  class ARROW_EXPORT SyncEvent {
+   public:
+    virtual ~SyncEvent() = default;
+
+    /// @brief Block until sync event is completed.
+    ///
+    /// Should be a no-op for CPU devices.
```

Comment:
Are we sure there's no CPU paradigm where things could be asynchronous?

Reply:
I think the assumption here is that if you're dealing with asynchronicity with a CPU, you would manage that via threads/mutexes rather than using this paradigm directly. But i'm open to removing that statement if you think it would be better to leave that open for future development to not be restricted with that guarantee.

Reply:
> But i'm open to removing that statement if you think it would be better to leave that open for future development to not be restricted with that guarantee.

Yea, I don't see a reason to dictate that it's a no-op for CPU devices. If it's a no-op that's an implementation detail.